pkg/controller/disruption/disruption.go

```diff
@@ -311,11 +311,13 @@ func (r *ReconcileMachineDisruption) updateMachineDisruptionBudgetStatus(
 	expectedCount int32,
 	disruptedMachines map[string]metav1.Time) error {
 
+	// we add one because we do not want to respect disruption budget when expected and healthy are equal
+	disruptionsAllowed := currentHealthy - desiredHealthy + 1
```

Comment:
this does not seems related to the title of this commit

Reply:
@enxebre  I encounter the problem with it when I tested integration, can I just add a better description under the commit instead of creating a new one?

Reply:
Question to this particular line. Does it mean that when there is at least the same number of currently healthy machines as desired machines, we will no allow mhc to fix the remaining unhealthy machines? `isMachineDisruptionAllowed` has the second condition pretty clear:

```go
if mdb.Status.MachineDisruptionsAllowed < 0 {
	glog.Warningf("the machine disruption budget %s MachineDisruptionsAllowed is negative", mdb.Name)
	return false
}
```

I still want the MHC to be able to re-mediate machines even when `currentHealthy > desiredHealthy`.

Reply:
> we add one because we do not want to respect disruption budget when expected and healthy are equal

I don't understand this ^^ expected=desired, healthy=current? I'm confused

Also if  `disruptionsAllowed := currentHealthy - desiredHealthy`
if currentHealthy=3 and desiredHealthy=5 why wouldn't that allow disruptions??
This needs to clarified  and cleaned along with `getExpectedMachineCount` and `getExpectedScale which names are confusing
Also why `getExpectedScale` look at the Replicas at all, that's pretty odd since we are talking about machines here, how does replicas comes in to the game?

Reply:
> Also if `disruptionsAllowed := currentHealthy - desiredHealthy`
> if currentHealthy=3 and desiredHealthy=5 why wouldn't that allow disruptions?? naming is very confusing.

When `currentHealthy` < `desiredHealthy` it means that you do not have enough quorum to remediate machines, as a good example you can see CEPH quorum, you do not want to delete unhealthy machines until you will have a quorum again, because you do not really know if some connectivity problem or maybe data corrupted on machines, so we want to prevent remediation in this case until we will have quorum again.

> This needs to clarified and cleaned along with `getExpectedMachineCount` and `getExpectedScale` which names are confusing

I can agree that `expectedCount` does not suit well, maybe something like `totalMachinesNumber` will suit much better, but regarding all other names, I think it explains well the purpose of the method.

> Also why `getExpectedScale` look at the Replicas at all, that's pretty odd since we are talking about machines here, how does replicas comes in to the game?

To calculate the total number of machines and number of desired healthy machines we need to get `replicas` from the owner machine controller(if it has some), because of it possible that we have `machineset` with replica 2, but it has only one machine object because `machineset`  failed to create the second machine from some reason, but we want to take this machine into account when we calculate the `desiredHealthy` parameter.
```
if mdb.Spec.MaxUnavailable != nil {
		expectedCount = r.getExpectedScale(mdb, machines)
		desiredHealthy = expectedCount - int32(*mdb.Spec.MaxUnavailable)
		if desiredHealthy < 0 {
			desiredHealthy = 0
		}
	} else if mdb.Spec.MinAvailable != nil {
		desiredHealthy = *mdb.Spec.MinAvailable
		expectedCount = int32(len(machines))
	}
```

> Also this seems to be relying on `IsMachineHealthy` which use static conditions which might or might not be the same than the ones defined as the node-problem

Yep it should be fixed, will do it under the separate PR.

Reply:
> When currentHealthy < desiredHealthy it means that you do not have enough quorum to remediate machines, as a good example you can see CEPH quorum, you do not want to delete unhealthy machines until you will have a quorum again, because you do not really know if some connectivity problem or maybe data corrupted on machines, so we want to prevent remediation in this case until we will have quorum again.

That line is inside `updateMachineDisruptionBudgetStatus` method. So, I still don't see why the decision about respecting or not respecting the disruption budget should be made inside this method. If you want to have the MDB signal the MHC can not delete any additional machine since the disruption budget would be violated otherwise, the only way to do it is to set `mdb.Status.MachineDisruptionsAllowed` to zero.

Reply:
> Also if disruptionsAllowed := currentHealthy - desiredHealthy
if currentHealthy=3 and desiredHealthy=5 why wouldn't that allow disruptions?? naming is very confusing. This needs to clarified and cleaned along with getExpectedMachineCount and getExpectedScale which names are confusing

@enxebre's point is valid. If you want to make sure you will not start healing machines when the quorum is not reached, you set `MaxUnavailable` to be 1 or 2. That's not something that can be handled on the code level.

Reply:
If any, computation of `disruptionsAllowed` value does not belong inside `updateMachineDisruptionBudgetStatus`. `reconcile` itself might be the good place for it. Possibly inside another local method that will compute the value and provide justification why it's done this way.

Reply:
Just to be clear, I am commenting on the `+1` case. If the disruption budget is violated (currentlyHealthy < desiredHealthy), no disruption is allowed as expected.

Reply:
> `// we add one because we do not want to respect disruption budget when expected and healthy are equal`

If you allow to disrupt another machine when `currentHealthy == desiredHealthy`, you will violate the disruption budget. That's not right. Can you undo the change and open another PR we can discuss this particular case?